src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -72,37 +75,36 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
     try!(links::validate(deps));
 
     let dest = if build_config.release {"release"} else {"debug"};
-    let root = if resolve.root() == pkg.package_id() {
-        pkg
-    } else {
-        deps.iter().find(|p| p.package_id() == resolve.root()).unwrap()
-    };
+    let root = deps.iter().find(|p| p.package_id() == resolve.root()).unwrap();
     let host_layout = Layout::new(config, root, None, &dest);
     let target_layout = build_config.requested_target.as_ref().map(|target| {
         layout::Layout::new(config, root, Some(&target), &dest)
     });
 
     let mut cx = try!(Context::new(resolve, sources, deps, config,
-                                   host_layout, target_layout, pkg,
+                                   host_layout, target_layout, root,
                                    build_config, profiles));
 
     let mut queue = JobQueue::new(cx.resolve, deps, cx.jobs());
 
-    // Prep the context's build requirements and see the job graph for all
-    // packages initially.
     {
         let _p = profile::start("preparing build directories");
```

Comment:
This profile marker isn't quite profiling the same thing any more, could you be sure to keep the scope the same as it was before?